narwhals/translate.py

```diff
@@ -294,131 +304,64 @@ def func(df):
     def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
         @wraps(func)
         def wrapper(*args: Any, **kwargs: Any) -> Any:
-            if args:
-                df_any = args[0]
-            elif kwargs:
-                params = list(inspect.signature(func).parameters.keys())
-                first_key = params[0]
-                df_any = kwargs[first_key]
-            else:
-                raise TypeError("Expected function which takes at least one argument.")
-            df = from_native(
-                df_any,
-                strict=strict,
-                eager_only=eager_only,
-                series_only=series_only,
-                allow_series=allow_series,
-            )
-            if args:
-                result = func(df, *args[1:], **kwargs)
-            else:
-                kwargs[first_key] = df
-                result = func(**kwargs)
+            args = [
+                from_native(
+                    arg,
+                    strict=strict,
+                    eager_only=eager_only,
+                    series_only=series_only,
+                    allow_series=allow_series,
+                )
+                for arg in args
+            ]  # type: ignore[assignment]
+
+            kwargs = {
+                name: from_native(
+                    value,
+                    strict=strict,
+                    eager_only=eager_only,
+                    series_only=series_only,
+                    allow_series=allow_series,
+                )
+                for name, value in kwargs.items()
+            }
```

Comment:
@MarcoGorelli should we check that passed dataframe/series are coming from the same backend?
It may be not so trivial, but on the other hand, it could speed up parsing by avoiding loading backend libraries N times and some number of checks.

Reply:
I think we have a check in `join` that objects are from the same library

not sure if it's worth it though, maybe it can be removed, and we just assume it's all from the same library?

Reply:
> not sure if it's worth it though, maybe it can be removed, and we just assume it's all from the same library?

I would be fine with giving responsibility to the users for this. It is also significantly easier to validate once one knows which arguments are dataframe/series, rather than dataframe/series or anything else.

> I think we have a check in join that objects are from the same library

I think it is significantly more complex as:

- non-dataframe elements can be passed
- once the first dataframe is found, need to check all other arguments to have the same backend, but only if dataframes, otherwise just pass them along

I didn't put much time on this yet though. I will give it a try soon-ish 👌

Reply:
right, thanks - if anyone's mixing different libraries together they'll likely need to do some data conversion anyway, so narwhals probably isn't the library for their use case anyway, OK to just document the assumption

Reply:
One way of doing it is:

- parse everything (as current implementation), then
- validate backends

```py
backends = {
    b() for v in [*args, *kwargs.values()] if (b:=getattr(v, "__native_namespace__", None))
}

if len(backends) > 1:
    msg = "Found multiple backends. Make sure that all inputs come from the same one."
    raise ValueError(msg)
```

The issue I have with this is:

- It adds some overhead
- It fails with Series, as they do not implement `__native_namespace__`

WDYT?

Reply:
I think this should be fine - the overhead should be minimal, as we'd just be loading modules which the user has already imported. We can also add `__native_namespace__` to `Series`

We can time it, but my intuition is that it's probably low enough to be acceptable

Reply:
Just pushed such change, which includes adding `__native_namespace__` to `Series` 😎

Reply:
I like what you've done here